src/ringct/rctSigs.cpp

```diff
@@ -1578,6 +1581,68 @@ namespace rct {
       }
     }
 
+    static struct SRctCache
+    {
+      mutable std::mutex m;
+      std::unordered_set<crypto::hash> hashes;
+      crypto::hash buf[8192] = {};
+      uint32_t counter = 0;
+
+      template<typename T>
+      static crypto::hash get_hash(const T& data)
+      {
+        std::stringstream ss;
+        binary_archive<true> ar(ss);
+        ::do_serialize(ar, const_cast<T&>(data));
+
+        crypto::hash h;
+        cryptonote::get_blob_hash(ss.str(), h);
+
+        return h;
+      }
+
+      void add(const crypto::hash& h)
+      {
+        std::lock_guard<std::mutex> lock(m);
+        if (hashes.insert(h).second)
+        {
+          crypto::hash& old_hash = buf[counter++ % 8192];
+          hashes.erase(old_hash);
+          old_hash = h;
+        }
+      }
+
+      bool find(const crypto::hash& h) const
+      {
+        std::lock_guard<std::mutex> lock(m);
+        return (hashes.find(h) != hashes.end());
+      }
+    } rct_cache;
+
+    bool verRctNonSemanticsSimple(const rctSig & rv)
+    {
+      // Hello future Monero dev! If you got this assert, read the following carefully:
+      //
+      // RCT cache assumes that SRctCache::get_hash will serialize and hash all rv's fields used for RingCT verification
+      // If you're about to add a new RCTType here, first you must check that binary_archive serialization writes all rv's fields to the binary blob
+      // If it's not the case, rewrite SRctCache::get_hash to serialize everything, even some "temporary" fields which are not serialized normally
+      CHECK_AND_ASSERT_MES_L1(rv.type <= RCTTypeBulletproofPlus, false, "Unknown RCT type. Make sure RCT cache works correctly with this type and then enable it in the code here.");
+
+      // Don't cache older (or newer) rctSig types
+      if (rv.type != RCTTypeCLSAG && rv.type != RCTTypeBulletproofPlus)
+        return verRctNonSemanticsSimple_internal(rv);
+
+      const crypto::hash h = SRctCache::get_hash(rv);
```

Comment:
Regardless of all that, `verRctNonSemanticsSimple` will return the same result for the same rctSigs. The cache doesn't change this behavior in such cases. Also, as you said, key image checks prevent that.

Reply:
Except that only works if this is after the images are written into the sigs (which it is) AND the sigs serialize them.

```c++
    // CLSAG signature
    struct clsag {
        keyV s; // scalars
        key c1;

        key I; // signing key image
        key D; // commitment key image

        BEGIN_SERIALIZE_OBJECT()
            FIELD(s)
            FIELD(c1)
            // FIELD(I) - not serialized, it can be reconstructed
            FIELD(D)
        END_SERIALIZE()
    };
```

CLSAG doesn't serialize its image.

I'm pretty sure a block can have TX0 with whatever, and then just reuse those RCT Sigs in TX1, unless there's a second serialization function in use which I'm missing.

Reply:
It looks like unit tests didn't catch this. MLSAG has the same "can be reconstructed" field and unit tests fail with cache if I enable it for MLSAG. I just checked the code and `RCTTypeBulletproofPlus` serializes everything, so I'll change my code to only cache `RCTTypeBulletproofPlus`. We don't have new CLSAG transactions anymore, right?

Reply:
And also `message` and `mixRing` fields need to be serialized separately: https://github.com/monero-project/monero/blob/master/src/ringct/rctTypes.h#L330
I'll fix the code.

Reply:
Pushed an update with more comments, `message` and `mixRing` serialized, and only `RCTTypeBulletproofPlus` cached.

Reply:
> absolute mixRing

What is it? What I added is `mixRing` field which is a vector of vector of pairs of 32-byte keys.

Reply:
Yep. I was using the adjective absolute on it since it's no longer relative (as key offsets is). No further commentary there :)